core/src/main/java/hudson/PluginManager.java

```diff
@@ -705,13 +705,14 @@ public boolean accept(File dir, String name) {
                     // If this was a plugin that was detached some time in the past i.e. not just one of the
                     // plugins that was bundled "for fun".
                     if (ClassicPluginStrategy.isDetachedPlugin(name)) {
-                        // If it's already installed and the installed version is older
-                        // than the bundled version, then we upgrade. The bundled version is the min required version
-                        // for "this" version of Jenkins, so we must upgrade.
                         VersionNumber installedVersion = getPluginVersion(rootDir, name);
                         VersionNumber bundledVersion = getPluginVersion(dir, name);
-                        if (installedVersion != null && bundledVersion != null && installedVersion.isOlderThan(bundledVersion)) {
-                            return true;
+                        // If the plugin is already installed, we need to decide whether to replace it with the bundled version.
+                        if (installedVersion != null && bundledVersion != null) {
+                            // If the installed version is older than the bundled version, then it MUST be upgraded.
```

Comment:
I'm not convinced this is correct (although my best guess is that this is a good idea). I thought a change that @stephenc did a while back did this, but it was something else, and I vaguely remember a discussion with @jglick about whether this would be desirable. I hope he has better memory than me and an opinion on this.

Reply:
Just to clarify, are you talking about the first case in particular (installed older than detached), or the change in general? 

Could it be something related to [this comment](https://github.com/dwnusbaum/jenkins/blob/14709ed2114790177e146efb0aa54f3cb4f9c0da/core/src/main/java/hudson/PluginManager.java#L737) or maybe [JENKINS-37041](https://issues.jenkins-ci.org/browse/JENKINS-37041)

Reply:
I think @daniel-beck refers to 6ea7883fed497b68eec9f89ec89082edf1b39ce5, which is in the other side of the  `if` (i.e., not an upgrade)

Reply:
@andresrc Yes, that's the commit. It didn't do what I thought it did, and I vaguely remember a conversation with @jglick about it.

Reply:
@dwnusbaum has much more recent context than I about everything here.

Reply:
AIUI the other side of the `if` (added in 6ea7883) ensures that any detached plugin that is already installed is at least the version bundled in the current war.

I am not sure about the reason for [this comment](https://github.com/dwnusbaum/jenkins/blob/14709ed2114790177e146efb0aa54f3cb4f9c0da/core/src/main/java/hudson/PluginManager.java#L737) regarding that code, but one thing that stands out to me is that dependencies of detached plugins are bundled in the war, but will not be listed in `ClassicPluginStrategy#getDetachedPlugins`, so maybe the concern is that if the detached plugin itself is up-to-date, but not its dependency, Jenkins could fail to start because the dependency is outdated?